### DISCUSSION TITLE
Synchronizing two Liquidsoap instances?

### DISCUSSION THREAD

**Discussion by codeofdusk** [savonet/liquidsoap#2400] [Q&A]
I have two machines with identical copies of a large music directory. Since both machines have the files locally, I'd prefer not to waste bandwidth by streaming the audio itself. However, I'd like both to play the same track at the same time (respecting source skips, etc.). How might I accomplish this? Writing something on one of the machines to receive the current track from the other via some sort of custom `request.dynamic`?

**Comment by toots** [comment]
Hi! You should be able to send commands form one liq to the other ones. You can register custom telnet and http handlers, see: https://www.liquidsoap.info/doc-2.0.4/server.html and https://www.liquidsoap.info/doc-2.0.4/harbor_http.html

**Reply by codeofdusk** [reply]
Thanks @toots! I've written this small script to receive commands to play on a secondary instance from the primary:

```
set("server.telnet", true)
set("server.telnet.port", 1234)

security = sine(1004.)

player = request.player(simultaneous=false)
server.register(
usage="play <song>",
description="Play a new song instantly.",
"play",
fun(uri) -> begin player.play(request.create(uri)) "Done!" end
)

radio = fallback(track_sensitive=false, [player, security])
output.portaudio(radio)
```

I'm assuming that I should add an `on_track` handler to the primary server's script to connect to the new second instance and send `play <local path>`. But I'm not sure how to do this.

I'm also thinking that it might be simpler to reverse the paradigm and have the new, secondary instance connect to the primary to receive tracks to play in a longer lived connection, rather than opening a new connection for each track change? (the primary is facing the public internet, secondary is behind a firewall/NAT).
